# Project: Online Retail Exploratory Data Analysis with Python

## Overview

In this project, I demonstrate entry-level data analytics in an online retail company, helping to interpret real-world data to help make a key business decision.

## Case Study
In this project, we worked with transactional data from an online retail store. The dataset contains information about customer purchases, including product details, quantities, prices, and timestamps. In this project, an exploration and analysis of this dataset was performed to obtain information about the store's sales trends, customer behavior, and most popular products. 

By conducting exploratory data analysis, we identified patterns, outliers and correlations in the data, which will allow us to make data-driven decisions and recommendations to optimize store operations and improve customer satisfaction. Through visualizations and statistical analysis, you will uncover key trends, such as the store's best-selling months, best-selling products, and most valuable customers. Ultimately, this project aims to provide actionable information that can drive strategic business decisions and improve overall store performance in the competitive online retail marketplace.


## Info
The python version of this project is: 3.12.4
The packages used for this project were:

- pandas
- numpy
- seaborn
- matplotlib

## Project Objectives
1. Describe data to answer key questions to uncover insights
2. Gain valuable insights that will help improve online retail performance
3. Provide analytic insights and data-driven recommendations

## Dataset

The dataset worked with is the “Online Retail” dataset. It contains transactional data from an online retail store from 2010 to 2011. The dataset is available as an .xlsx file named `Online Retail.xlsx`. 

The dataset contains the following columns

- InvoiceNo: Invoice number of the transaction
- StockCode: Unique product code
- Description: Product description: Description of the product
- Quantity: Quantity of the product in the transaction
- InvoiceDate: Date and time of the transaction
- UnitPrice: Unit price of the product
- CustomerID: Unique customer identifier
- Country: Country in which the transaction took place

In [1]:
import pandas as pd
import matplotlib as plt
import numpy as np
import seaborn as sns

In [2]:
df=pd.read_excel("Online Retail.xlsx")

In [3]:
print(df.head(10))

  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26

In [4]:
df.shape

(541909, 8)

In [5]:
print(df.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB
None


Let's look at the types, and how the columns _'Description'_ and _'CustomerID'_ have less than 541909 unfilled data.

Now that we know a little more about our data set, let's fill in those blanks and remove duplicate data.

In [6]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Now we know the amount of data that are null, this can happen either by human error or technical failure, decreasing the quality of the dataset.

As the corresponding columns represent information from a sale, we cannot delete them completely, so we will fill in these blanks.

In [7]:
df.fillna({"Description": "NA"}, inplace=True)
df.fillna({"CustomerID": 0.0}, inplace=True)

We will check again if the task has been accomplished

In [8]:
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

Once those blanks are filled in, what data will be duplicated? Let's see:

In [9]:
df.duplicated().sum()

5268

Now let's remove the duplicates and check again how many duplicates there are and if the dataset size decreased.

In [10]:
df.drop_duplicates(inplace = True)

In [11]:
df.duplicated().sum()

0

In [12]:
df.shape

(536641, 8)

Our dataset no longer has unfilled spaces, nor duplicates, now we can start exploring the dataset.

In [13]:
df.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  536641.000000                         536641  536641.000000   
mean        9.620029  2011-07-04 08:57:06.087421952       4.632656   
min    -80995.000000            2010-12-01 08:26:00  -11062.060000   
25%         1.000000            2011-03-28 10:52:00       1.250000   
50%         3.000000            2011-07-19 14:04:00       2.080000   
75%        10.000000            2011-10-18 17:05:00       4.130000   
max     80995.000000            2011-12-09 12:50:00   38970.000000   
std       219.130156                            NaN      97.233118   

          CustomerID  
count  536641.000000  
mean    11435.904653  
min         0.000000  
25%         0.000000  
50%     14336.000000  
75%     16241.000000  
max     18287.000000  
std      6795.044250

For this case, the CustomerID information is not relevant, since there are data that are not representative, such as “0”, since they made the purchase but we do not have their CustomerID data.

In [14]:
df['CustomerID'] = df['CustomerID'].astype(str)

In [15]:
df.describe()

Quantity                    InvoiceDate      UnitPrice
count  536641.000000                         536641  536641.000000
mean        9.620029  2011-07-04 08:57:06.087421952       4.632656
min    -80995.000000            2010-12-01 08:26:00  -11062.060000
25%         1.000000            2011-03-28 10:52:00       1.250000
50%         3.000000            2011-07-19 14:04:00       2.080000
75%        10.000000            2011-10-18 17:05:00       4.130000
max     80995.000000            2011-12-09 12:50:00   38970.000000
std       219.130156                            NaN      97.233118

Note that there are negative values in Quantity and UnitPrice, this may represent a return, a discount, and that is why we will not eliminate them.
Also note that we do not have the total of the purchases made, for this, we create a Total Price column, and then we will organize it by InvoiceNo.

In [16]:
df['TotalPrice'] = df['Quantity'] * df['UnitPrice']
print(df.head())

  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  TotalPrice  
0 2010-12-01 08:26:00       2.55    17850.0  United Kingdom       15.30  
1 2010-12-01 08:26:00       3.39    17850.0  United Kingdom       20.34  
2 2010-12-01 08:26:00       2.75    17850.0  United Kingdom       22.00  
3 2010-12-01 08:26:00       3.39    17850.0  United Kingdom       20.34  
4 2010-12-01 08:26:00       3.39    17850.0  United Kingdom       20.34  


In [18]:
total_by_invoice = df.groupby('InvoiceNo')['TotalPrice'].sum().reset_index()
print(total_by_invoice)

      InvoiceNo  TotalPrice
0        536365      139.12
1        536366       22.20
2        536367      278.73
3        536368       70.05
4        536369       17.85
...         ...         ...
25895   C581484  -168469.60
25896   C581490      -32.53
25897   C581499     -224.69
25898   C581568      -54.75
25899   C581569       -7.50

[25900 rows x 2 columns]


In [17]:
df.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  536641.000000                         536641  536641.000000   
mean        9.620029  2011-07-04 08:57:06.087421952       4.632656   
min    -80995.000000            2010-12-01 08:26:00  -11062.060000   
25%         1.000000            2011-03-28 10:52:00       1.250000   
50%         3.000000            2011-07-19 14:04:00       2.080000   
75%        10.000000            2011-10-18 17:05:00       4.130000   
max     80995.000000            2011-12-09 12:50:00   38970.000000   
std       219.130156                            NaN      97.233118   

          TotalPrice  
count  536641.000000  
mean       18.123861  
min   -168469.600000  
25%         3.750000  
50%         9.870000  
75%        17.400000  
max    168469.600000  
std       380.656263

Since we observed some trends of central measurement and dispersion, we will calculate these and more, and make a more meaningful analysis.

In [32]:
print("mean:",df["Quantity"].mean(),"\n"
"median:",df["Quantity"].median(),"\n"
"mode:",df["Quantity"].mode(),"\n"
"range:",df["Quantity"].max()-df["Quantity"].min(),"\n"
"variance:",df["Quantity"].var(),"\n"
"standard deviation:",df["Quantity"].std())


mean: 9.620029032444409 
median: 3.0 
mode: 0    1
Name: Quantity, dtype: int64 
range: 161990 
variance: 48018.02532598476 
standard deviation: 219.13015613097335


In [33]:
print("mean:",df["UnitPrice"].mean(),"\n"
"median:",df["UnitPrice"].median(),"\n"
"mode:",df["UnitPrice"].mode(),"\n"
"range:",df["UnitPrice"].max()-df["UnitPrice"].min(),"\n"
"variance:",df["UnitPrice"].var(),"\n"
"standard deviation:",df["UnitPrice"].std())

mean: 4.632655674836623 
median: 2.08 
mode: 0    1.25
Name: UnitPrice, dtype: float64 
range: 50032.06 
variance: 9454.27916555312 
standard deviation: 97.23311763773246


In [34]:
print("mean:",df["TotalPrice"].mean(),"\n"
"median:",df["TotalPrice"].median(),"\n"
"mode:",df["TotalPrice"].mode(),"\n"
"range:",df["TotalPrice"].max()-df["TotalPrice"].min(),"\n"
"variance:",df["TotalPrice"].var(),"\n"
"standard deviation:",df["TotalPrice"].std())

mean: 18.123861117581395 
median: 9.870000000000001 
mode: 0    15.0
Name: TotalPrice, dtype: float64 
range: 336939.2 
variance: 144899.1906691031 
standard deviation: 380.6562631418313


In [ ]:
df_Dfiltered = df[df['Description'] != 'NA']

print(df_Dfiltered)

In [ ]:
frecuencia_Stockcode=df["StockCode"].value_counts()
print(frecuencia_Stockcode)

In [ ]:
frecuencia_Description=df["Description"].value_counts()
print(frecuencia_Description)

In [20]:
df_CIDfiltered = df[df['CustomerID'] != '0.0']

print(df_CIDfiltered)

       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  TotalPric

In [ ]:
frecuencia_CustomerID=df["CustomerID"].value_counts()
print(frecuencia_CustomerID)

In [ ]:
frecuencia_InvoiceDate=df["InvoiceDate"].value_counts()
print(frecuencia_InvoiceDate)

In [ ]:
frecuencia_Country=df["Country"].value_counts()
print(frecuencia_Country)

In [ ]:
df.head(5)

In [ ]:
x=df["InvoiceDate"]
sns.histplot(x)
plt

In [ ]:
frecuencias = df['Country'].value_counts()

frecuencias_df = frecuencias.reset_index()
frecuencias_df.columns = ['País', 'Frecuencia']

sns.barplot(x='País', y='Frecuencia', data=frecuencias_df)



# Rotar etiquetas si son largas
plt

In [ ]:
sns.boxplot(x='Country', y='TotalPrice', data=df)
plt

In [ ]:

df["Quantity"].plot()
df["UnitPrice"].plot()
plt

In [ ]:
df.plot(kind = 'scatter', x = 'InvoiceDate', y = 'UnitPrice')
plt

In [ ]:
df.plot(kind = 'scatter', x = 'InvoiceDate', y = 'Quantity')
plt

In [ ]:
df.plot(kind = 'scatter', x = 'InvoiceDate', y = 'TotalPrice')
plt

In [86]:
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

In [87]:
df['Month'] = df['InvoiceDate'].dt.month
df['DayOfWeek'] = df['InvoiceDate'].dt.day_name()

In [90]:
sales_by_month = df.groupby('Month')['TotalPrice'].sum()
sales_by_day = df.groupby('DayOfWeek')['TotalPrice'].sum()
ordered_days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
sales_by_day = sales_by_day.reindex(ordered_days)

In [ ]:
sales_by_month.plot(kind='bar')
plt

In [ ]:
sns.barplot(x=sales_by_day.index, y=sales_by_day.values)

In [ ]:
top_products = df.groupby('Description')['Quantity'].sum().sort_values(ascending=False)

print(top_products.head(10))

In [ ]:
top_countries = df.groupby('Country')['Quantity'].sum().sort_values(ascending=False)

print(top_countries.head(10))

In [ ]:
top_products.head(10).plot(kind='bar')
plt

In [ ]:
top_countries.head(10).plot(kind='bar')
plt

In [ ]:
sns.boxplot(df['Quantity'])

In [ ]:
sns.boxplot(df['UnitPrice'])

In [ ]:
sns.scatterplot(x='Quantity', y='UnitPrice', data=df)

In [ ]:
valores_negativos = df[(df['Quantity'] < 0) | (df['UnitPrice'] < 0)]
print(valores_negativos)

In [ ]:
def find_outliers_iqr(data, column):
    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = data[(data[column] < lower_bound) | (data[column] > upper_bound)]
    return outliers

# Identify outliers in Quantity
outliers_quantity = find_outliers_iqr(df, 'Quantity')
print("Outliers in Quantity:")
print(outliers_quantity)

# Identify outliers in UnitPrice
outliers_price = find_outliers_iqr(df, 'UnitPrice')
print("Outliers in UnitPrice:")
print(outliers_price)

In [101]:
df_clean = df[(df['Quantity'] > 0) & (df['UnitPrice'] > 0)]


In [ ]:
df_clean

In [ ]:
df_clean.plot(kind = 'scatter', x = 'InvoiceDate', y = 'UnitPrice')
plt

In [ ]:
sns.scatterplot(x='Quantity', y='UnitPrice', data=df_clean)